### DISCUSSION TITLE
How to properly recreate DynamicTree after deserialization

### DISCUSSION THREAD

**Discussion by opera-aberglund** [louis-langholtz/PlayRho#578] [General]
HI again!

I ended up trying a different approach for serialization/deserialization, because I kept having problems recreating the state perfectly without desyncs. I wanted to make PlayRho innately serializable and compatible with Cista, such that you can simply call `cista::serialize(world)` directly, which also makes it more performant. For that I had to convert all the inner types to Cista's types, like `cista::vector` and `cista::unique_ptr` instead of the standard library ones. You can look at my implementation in the fork of PlayRho among my repositories, and [here](https://github.com/opera-aberglund/PlayRho/blob/e505710de104318166784418e6b1080f39e929c8/Library/include/playrho/d2/AabbTreeWorld.hpp#L585) you can see the members I declared for Cista to include in the serialization. It got a bit ugly because I had to get rid of the "Concept" virtual classes because they are not serializable, so the `World` object now simply points directly to a `WorldModel`, and similarly for the shapes and joints which now use a variant for the different types.

This works, but now I want to try to reduce the state size. Mainly, I'm looking at how I can avoid to serialize the dynamic tree. I ran into issues when I didn't serialize it, and I suspect it has to do with it being out of sync with the caches/proxies. I was wondering if you could provide some insight here? Previously I only serialized the bodies, shapes and joints mostly and recreated the world with the `CreateBody`, `CreateShape` etc. functions, but now the objects will be recreated directly without calling these functions. Thus, the tree is not affected until i run the Step function. It would be nice if there was some way to correctly rebuild the tree after deserialization of the other world members.

**Comment by louis-langholtz** [comment]
Hi there! Thank you for this update. I will try and look into this and get back with you with what I can. It sounds concerning & surprising to me that one might find working around the `Concept` polymorphic inner types a way to go. It's also disappointing to me that the dynamic tree would need to be serialized. Ideally to me, it's just a mechanism for optimizing simulations.

**Reply by opera-aberglund** [reply]
I actually had a discussion [here](https://github.com/felixguendling/cista/issues/231) about it with the creator of Cista and it didn't seem possible unfortunately. He recommended using variants instead of polymorphism. 

About the dynamic tree, maybe you could run a test that clears the entire tree before running a step, because that's essentially what's happening here if I serialize without it. 

Also, I thought about if it would be possible to do this whole serialization thing without changing any of the existing code, but instead just providing another implementation of the world concept that supports serialization. I felt a bit uneasy changing the core structure of the engine, so that would certainly be more elegant.

**Reply by louis-langholtz** [reply]
I am not familiar with Cista (or at least don't recall looking at it before). I took a look this evening at it though from the link of your discussion.

As serialization/deserialization approaches go, I'd stay away from using an intrusive mechanism and use a non-intrusive mechanism instead. Fortunately, it appears Cista also supports non-intrusive mechanisms (in addition to an intrusive mechanism).

I.e., I'd stay away from adding `cista_members()` functions to existing classes. Instead, I'd "override the serialize and deserialize functions" (from [here](https://github.com/felixguendling/cista/wiki/Custom-(De-)Serialization-Functions)). Details are available for doing that for [serialization](https://github.com/felixguendling/cista/wiki/Custom-(De-)Serialization-Functions#serialization) and [deserialization](https://github.com/felixguendling/cista/wiki/Custom-(De-)Serialization-Functions#deserialization).

A problem in my opinion with intrusive approaches to serialization is a bias towards the innards of classes (like private member variables) instead of the interfaces of classes (like publicly accessible functions or variables). Another problem with intrusive approaches, is tighter coupling.

In your fork of PlayRho, it looked to me like you took the intrusive approach. Did you consider writing custom Cista serialization & deserialization functions for the PlayRho types instead? If so, what steered you away from that approach to adding the intrusive `cista_members()` member functions?

Anyways, writing custom serialization and deserialization functions and coming at that task from an "outside" of the encapsulating surface area perspective, aligns more preferably to me to PlayRho's design and is hopefully easier. If there's not access to something, it's because I didn't see that something as necessary for "equational completeness". Or, said another way, I didn't see that something as necessary for serialization/deserialization.

That brings me to the dynamic tree.

You may be correct that with the existing surface areas that PlayRho types are providing one needs to also serialize/deserialize the dynamic tree. In my opinion, that'd be a bug in PlayRho however as that'd violate the value semantics and "equational completeness" of the library's types.

Relatedly, I still forget sometimes that `Contact` objects in PlayRho - nor Box2D - aren't actually contacts of two shapes of two bodies. They're overlaps of those shape's movement extended axis aligned bounding boxes (AABBs). I point this out because I'm wanting to make sure is that what you're calling _desyncs_, isn't only seeing different `Contact` objects but specifically seeing different simulation results in terms of bodies and/or collisions. If the latter, then I need to figure a different way to do the dynamic tree that doesn't depend on any properties that would need to be serialized/deserialized and without loosing performance.

Regarding "using variants instead of polymorphism" it seems to me that's more from a perspective of solving the problem of how to serialize pointers to base types of objects that are actually derived types. In other words, don't try to serialize/deserialize those pointers.

Instead of serializing pointer to base types of derived types, use the `GetType` and `TypeCast` interfaces to `Joint` and `Shape` types to get access to their underlying values as the types they are and then serialize those. Admittedly, if-trees are not as appealing to me as say `switch`-statements or interfaces like `std::visit`, and probably aren't as performant either. But it's a trade off for having types like `Joint` that a user could write a custom supporting type for without having to intrusively change `Joint` to support (like using `std::variant` for instance would). And one really only needs to do this in limited cases, like on serialization. Incidentally, I believe using `std::any` has the same trade off but wouldn't then place compile-time requirements on what a usable `Joint` type has to support.

**Reply by opera-aberglund** [reply]
I did choose the other non-intrusive approach first, it was just that I couldn't get it to work. I ran tests where I checked that the deserialized world was equal to the copied world through the equality operator, yet they still diverged after a few steps, leading to bodies having different positions and velocities. My guess is that there is something that the equality operator doesn't compare that is relevant for the future state.

I ended up with this approach because I wanted to try if just serializing everything agnostically would solve the issue, and it did, I just got a very large state because of the whole dynamic tree.

I could probably at least get back the concept classes as you say by using a custom serialization function just for those.

**Reply by opera-aberglund** [reply]
But you are saying now that the desyncs I've seen could be because of the dynamic tree, and that I just have to serialize it as it is now? Because I do would prefer to keep it the old way, just using public methods and not have to modify anything in the engine..

**Reply by opera-aberglund** [reply]
What exactly do you think the information is that is contained in the dynamic tree that has to be serialized? I'm thinking of how to work around it. I can't really serialize the entire tree because it's way too large, but I also need to find a way to maintain determinism.

**Reply by louis-langholtz** [reply]
> I did choose the other non-intrusive approach first

Glad to hear that!

> it was just that I couldn't get it to work.

Bummer to hear that.

Sounds like from that approach your code couldn't get to some state that it needs to serialize. Was that the dynamic tree then? There's the function `const DynamicTree& GetTree(const AabbTreeWorld& world) noexcept` that provides access to it though. So I feel a bit confused about this.

Admittedly, it'd help me if I looked through your fork more closely. Time for that is in short supply right now for me however.

>  I ran tests where I checked that the deserialized world was equal to the copied world through the equality operator, yet they still diverged after a few steps, leading to bodies having different positions and velocities.

Besides switching to the intrusive approach, what else changed in your implementation between the two? In the intrusive approach did you for example serialize different state or more state? I ask because this is essentially new ground for me (serializing the world and then deserializing it) and I'm hoping to gain the benefit of what you've found so far. Something you'd said suggested to me that you had to serialize the dynamic tree state but then something else suggested to didn't go that far (so, I'm unclear about this).

> But you are saying now that the desyncs I've seen could be because of the dynamic tree

What I was saying was that I'd prefer the dynamic tree being such that it wouldn't need to be serialized. If what I understand you to be saying is correct however, it'd seem like some state, possibly the dynamic tree, isn't working as I'd prefer then. And, I'll be wanting to address that starting with some things like:

1. Recognizing what other state needs to be serialized.
2. Understanding why that state needs to be serialized if it's not clear.
3. Considering whether that's a design issue or a configuration issue.

It sounds like you're saying that you suspect it's the dynamic tree state which needs to be serialized. And, put another way, if that's indeed the case I'm wanting to understand whether configuration settings like setting the `StepConf::aabbExtension` variable to the value of zero, would make a difference in this.

> that I just have to serialize it as it is now?

Not saying this, no.

> What exactly do you think the information is that is contained in the dynamic tree that has to be serialized? I'm thinking of how to work around it. I can't really serialize the entire tree because it's way too large, but I also need to find a way to maintain determinism.

I don't know. If I can easily identify that I will. I'd rather change the dynamic tree code though so it didn't need to be serialized to begin with. Sounds like we'd both prefer that. :)

**Reply by opera-aberglund** [reply]
Hey! Sorry for not replying in a while. What I actually ended up with was going back to use my non-intrusive way of serializing with not changing anything in PlayRho, and to solve the desync issue with the dynamic tree, I simply forced all players in the game to clear their dynamic tree cache and rebuild it each time a player joins the game. A bit suboptimal performance wise, but it guarantees that everyone has the same caches at all times and it worked surprisingly well in the game.

On another note, we just release the game "Timmersdala" we've been making with PlayRho, you can check out a blogpost from the GameMaker guys [here](https://gamemaker.io/en/blog/gamemaker-future-multiplayer), and test the game [here](https://gx.games/games/6syokf/timmersdala/) if you're curious! 

I'm also looking at if we can open source my serialization library, I'll update you on that.
